R/kenwardroger.R

```diff
@@ -9,98 +11,118 @@
 #' so they are not stacked in the result.
+#'
 #' @return Named list with elements:
-#' - `P`: `matrix` of P component.
-#' - `Q`: `matrix` of Q component.
-#' - `R`: `matrix` of R component.
+#' - `P`: `matrix` of \eqn{P} component.
+#' - `Q`: `matrix` of \eqn{Q} component.
+#' - `R`: `matrix` of \eqn{R} component.
+#'
 #' @keywords internal
 h_get_kr_comp <- function(tmb_data, theta) {
   .Call(`_mmrm_get_pqr`, PACKAGE = "mmrm", tmb_data, theta)
 }
 
 
 #' Calculation of Kenward-Roger Degrees of Freedom for Multi-Dimensional Contrast
+#'
 #' @param object `mmrm` object
 #' @param contrast `matrix` contrast matrix
 #' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#'
 #' @return List with `num_df`, `denom_df`, `f_stat` and `p_val` (2-sided p-value).
+#'
 #' @keywords internal
-df_md_kr <- function(object, contrast, linear = FALSE) {
-  if (object$tmb_data$reml != 1) {
-    stop("Kenward-Roger is only for REML!")
+h_df_md_kr <- function(object, contrast, linear = FALSE) {
+  if (component(object, "reml") != 1) {
+    stop("Kenward-Roger is only for REML")
   }
   kr_comp <- object$kr_comp
-  w <- solve(object$tmb_obj$he(object$theta_est))
+  w <- component(object, "theta_vcov")
   v_adj <- object$beta_vcov_adj
-  df <- h_kr_df(object$beta_vcov, contrast, w, kr_comp$P)
-  vsolve <- solve(contrast %*% v_adj %*% t(contrast))
-  f_statistic <- 1 / nrow(contrast) * object$beta_est %*% t(contrast) %*% vsolve %*% contrast %*% object$beta_est
+  df <- h_kr_df(v0 = object$beta_vcov, l = contrast, w = w, p = kr_comp$P)
+  vsolve <- solve(h_quad_form_mat(contrast, v_adj))
+  beta_contrast <- component(object, "beta_est") %*% t(contrast)
+  f_statistic <- 1 / nrow(contrast) * h_quad_form_mat(beta_contrast, vsolve)
   f_star <- f_statistic * df$lambda
-  ret <- list(
+  list(
     num_df = nrow(contrast),
     num_df = nrow(contrast),
     denom_df = df$m,
     f_stat = f_star[1, 1],
     p_val = pf(f_star[1, 1], nrow(contrast), df$m, lower.tail = FALSE)
   )
-  ret$sqrt <- sqrt(contrast %*% v_adj %*% t(contrast))
-  ret
 }
 
 #' Calculation of Kenward-Roger Degrees of Freedom for One-Dimensional Contrast
+#'
 #' @param object `mmrm` object
 #' @param contrast contrast vector
 #' @param linear `logical`\cr whether to use linear Kenward-Roger approximation
+#'
 #' @return List with `est`, `se`, `df`, `t_stat` and `p_val`.
+#'
 #' @keywords internal
-df_1d_kr <- function(object, contrast, linear = FALSE) {
-  if (object$tmb_data$reml != 1) {
+h_df_1d_kr <- function(object, contrast, linear = FALSE) {
+  if (component(object, "reml") != 1) {
     stop("Kenward-Roger is only for REML!")
   }
   assert_numeric(contrast, len = length(component(object, "beta_est")))
   est <- sum(contrast * component(object, "beta_est"))
-  kr_comp <- object$kr_comp
-  w <- solve(object$tmb_obj$he(object$theta_est))
+  w <- component(object, "theta_vcov")
```

Comment:
maybe nicer to avoid this temporary variable and instead just directly do this in the `h_kr_df()` call